clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -292,67 +370,49 @@ protected void onJoinComplete(int generation,
             this.nextAutoCommitTimer.updateAndReset(autoCommitIntervalMs);
 
         // execute the user's callback after rebalance
-        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
         switch (protocol) {
             case EAGER:
                 if (!ownedPartitions.isEmpty()) {
                     log.info("Coordinator has owned partitions {} that are not revoked with {} protocol, " +
                         "it is likely client is woken up before a previous pending rebalance completes its callback", ownedPartitions, protocol);
                 }
 
-                log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
-                try {
-                    listener.onPartitionsAssigned(assignedPartitions);
-                } catch (WakeupException | InterruptException e) {
-                    throw e;
-                } catch (Exception e) {
-                    log.error("User provided listener {} failed on partition assignment", listener.getClass().getName(), e);
-                }
+                // TODO: after refactored the error handling, we need to exclude owned partitions from assigned partitions
+
+                firstException.compareAndSet(null, maybeAssignPartitions(assignedPartitions));
+
                 break;
 
             case COOPERATIVE:
-                assignAndRevoke(listener, assignedPartitions, ownedPartitions);
+                Set<TopicPartition> addedPartitions = new HashSet<>(assignedPartitions);
+                Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
+                addedPartitions.removeAll(ownedPartitions);
+                revokedPartitions.removeAll(assignedPartitions);
+
+                log.info("Updating with newly assigned partitions: {}, compare with already owned partitions: {}, " +
+                        "newly added partitions: {}, revoking partitions: {}",
+                    Utils.join(assignedPartitions, ", "),
+                    Utils.join(ownedPartitions, ", "),
+                    Utils.join(addedPartitions, ", "),
+                    Utils.join(revokedPartitions, ", "));
+
+                // add partitions that was not previously owned but are now assigned
+                firstException.compareAndSet(null, maybeAssignPartitions(addedPartitions));
+
+                // revoked partitions that was previously owned but no longer assigned
+                firstException.compareAndSet(null, maybeRevokePartitions(tp -> !assignedPartitions.contains(tp)));
```

Comment:
I don't think this logic is quite right...when we call maybeRevokePartitions we calculate revokedPartitions = assignedPartitions.filter(tp -> !assignedPartitions.contains(tp)) which is an empty list.